### PR TITLE
Add InfluxDB customizer hook point

### PR DESCRIPTION
**Pull request by eddumelendez**
More configuration properties are added in order to allow
customization for `influxdb` bean. Also, in order to
customize `influxdb`, `InfluxDbCustomizer` is added.

See gh-20867

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbBatchOptionsCustomizer.java
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbBatchOptionsCustomizer.java:35]: That is an interesting idea but I am wondering why we don't provide a customizer for the high-level `InfluxDB`? If you need to customize something we might just as well give the user full control.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbAutoConfiguration.java
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbAutoConfiguration.java:59]: You can use `PropertyMapper` for this too. There is a way to invoke a method.

**Review by snicoll — CHANGES_REQUESTED** [review]
Thanks for the PR @eddumelendez. This looks good (I had no idea we could expose so many options). 

I think we need a `InfluxDbPropertiesTests` that validates the defaults are consistent (see other such *PropertiesTests for more details).

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbProperties.java
Comment (eddumelendez) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbProperties.java:164]: `flushDuration` and `jitterDuration` are set as integers with `TimeUnit.MILLISECONDS` so that the reason both are provided with such descriptions

See https://github.com/influxdata/influxdb-java/blob/master/src/main/java/org/influxdb/impl/InfluxDBImpl.java#L308

Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbProperties.java:164]: It doesn't matter. Most `Duration` we have in `ConfigurationProperties` are actually mapped to int/long values (check existing keys).

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbBatchOptionsCustomizer.java
Comment (eddumelendez) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbBatchOptionsCustomizer.java:35]: There is no other properties that can be set at `influxdb` level, just `query` or `write` operations. Also, there is no easy way to fetch `BatchOptions` to override those set by configuration properties with those provided by customizers. And, due to BatchOperations provide properties that can be set via lambdas then it was a much better option to provide a customizer for it.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbProperties.java
Comment (eddumelendez) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbProperties.java:164]: I had forgotten `@DurationUnit` 😅

Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbProperties.java:164]: DurationUnit has nothing to do with what we're discussing here IMO. I don't understand why you bring this up.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbBatchOptionsCustomizer.java
Comment (snicoll) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbBatchOptionsCustomizer.java:35]: > There is no other properties that can be set at influxdb level, just query or write operations.

Correct but that might change in the future and that doesn't change the fact that someone may want to customize something programmatically. Right now, you're offering that only for batch options while you could offer that for everything else. That doesn't seem legit to me.

> there is no easy way to fetch BatchOptions to override those set by configuration properties with those provided by customizers.

I know that but if you use the customizer you want to be in control. If the InfluxDB API does not offer a way to override part of the options, so be it.

**Comment by eddumelendez** [comment]
@snicoll thanks for the review and quickly replies. I will submit the changes later today.

**Comment by snicoll** [comment]
Thanks for the follow-up @eddumelendez. I've polished the PR but I want to hold off as reviewing this made me realize about the [new InfluxDB client](https://github.com/influxdata/influxdb-client-java). I think we need to investigate the changes before adding more properties here.

**Comment by snicoll** [comment]
We've discussed this and decided to only include the customizer for now. Looking at the new client, concepts have changed in such a way that adding more properties would mean we'd be in a tricky situation when upgrading to it. On the contrary, we could easily add support for the new client with the support that we have.

**Comment by bednar** [comment]
Hi @snicoll,

we want to prepare a support for the new driver.

What is a best way how to do it in terms of packaging? I think there are two possibility how to do it:

1. create new package `org.springframework.boot.autoconfigure.influx2`
2. use existing package `org.springframework.boot.autoconfigure.influx` and an implementation for new client will have naming such as: `InfluxDbClientAutoConfiguration`, `InfluxDbClientProperties`, ...

Or is there different approach?

Regards

**Comment by snicoll** [comment]
@bednar I've shared a commit in the comment just above yours, have you seen it?

**Comment by bednar** [comment]
Yeah I saw it 😏... I just want to be sure that we will select correct path. Thanks for quick reply 👍

**Comment by snicoll** [comment]
Not sure who is we. Given that I've already something, it would be better to get some feedback on that  rather than working on a separate PR (if that's what you want to do).

**Comment by bednar** [comment]
> Not sure who is we.

I'm maintainer of [influxdb-client-java](https://github.com/influxdata/influxdb-client-java)

> Given that I've already something, it would be better to get some feedback on that rather than working on a separate PR (if that's what you want to do).

I think that your implementation is fine (https://github.com/snicoll/spring-boot/tree/influx-client-java), just a preferred way for authentication is by `token` (https://github.com/snicoll/spring-boot/blob/0359550f4631ab1cdc41427ac83345438293f146/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/influx/InfluxDbClientAutoConfiguration.java#L52).

I can help you with tests and support for `HealthIndicator`.

**Comment by snicoll** [comment]
> I'm maintainer of influxdb-client-java

I am aware of that. I didn't understand what you mean by "we will select correct path".

> I think that your implementation is fine (https://github.com/snicoll/spring-boot/tree/influx-client-java), just a preferred way for authentication is by token

Thank you for reviewing. Let's move this conversation on ##25891. I know that token is preferred with v2 but that would introduce a v2 specific property and I am not sure yet we want to do that. I've flagged the issue for team meeting so that we can discuss our options.

**Pull request closed without merging**
